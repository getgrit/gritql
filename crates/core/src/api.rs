use crate::problem::{FileOwner, InputRanges, Problem};
use crate::{fs, tree_sitter_serde::tree_sitter_node_to_json};
use anyhow::{bail, Result};
use im::Vector;
use marzano_language::target_language::TargetLanguage;
use marzano_util::analysis_logs::AnalysisLog as MarzanoAnalysisLog;
use marzano_util::position::{Position, Range, VariableMatch};
use serde::{Deserialize, Serialize};
use serde_json::to_string_pretty;
use std::path::PathBuf;
use std::{fmt, ops::Range as StdRange, str::FromStr, vec};
use tree_sitter::Tree;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "__typename")]
pub enum MatchResult {
    #[serde(rename = "PatternInfo")]
    PatternInfo(PatternInfo),
    #[serde(rename = "AllDone")]
    AllDone(AllDone),
    #[serde(rename = "Match")]
    Match(Match),
    #[serde(rename = "InputFile")]
    InputFile(InputFile),
    #[serde(rename = "Rewrite")]
    Rewrite(Rewrite),
    #[serde(rename = "CreateFile")]
    CreateFile(CreateFile),
    #[serde(rename = "RemoveFile")]
    RemoveFile(RemoveFile),
    #[serde(rename = "DoneFile")]
    DoneFile(DoneFile),
    #[serde(rename = "AnalysisLog")]
    AnalysisLog(AnalysisLog),
}

/// Make a path look the way provolone expects it to
/// Removes leading "./", or the root path if it's provided
fn normalize_path_in_project<'a>(path: &'a str, root_path: Option<&'a PathBuf>) -> &'a str {
    #[cfg(debug_assertions)]
    if let Some(root_path) = root_path {
        if !root_path.to_str().unwrap_or_default().ends_with('/') {
            panic!("root_path must end with a slash.");
        }
    }
    if let Some(root_path) = root_path {
        let root_path = root_path.to_str().unwrap();
        path.strip_prefix(root_path).unwrap_or(path)
    } else {
        path.strip_prefix("./").unwrap_or(path)
    }
}

impl MatchResult {
    /// Make the paths compatible with provolone expectations, by removing leading "./"
    pub fn normalize_paths(&mut self, root_path: Option<&PathBuf>) {
        match self {
            MatchResult::PatternInfo(pi) => {
                pi.source_file = normalize_path_in_project(&pi.source_file, root_path).to_owned()
            }
            MatchResult::AllDone(_) => {}
            MatchResult::Match(m) => {
                m.source_file = normalize_path_in_project(&m.source_file, root_path).to_owned()
            }
            MatchResult::InputFile(input_file) => {
                input_file.source_file =
                    normalize_path_in_project(&input_file.source_file, root_path).to_owned()
            }
            MatchResult::Rewrite(r) => {
                r.original.source_file =
                    normalize_path_in_project(&r.original.source_file, root_path).to_owned();
                r.rewritten.source_file =
                    normalize_path_in_project(&r.rewritten.source_file, root_path).to_owned()
            }
            MatchResult::CreateFile(cf) => {
                cf.rewritten.source_file =
                    normalize_path_in_project(&cf.rewritten.source_file, root_path).to_owned()
            }
            MatchResult::RemoveFile(rf) => {
                rf.original.source_file =
                    normalize_path_in_project(&rf.original.source_file, root_path).to_owned()
            }
            MatchResult::DoneFile(df) => {
                df.relative_file_path =
                    normalize_path_in_project(&df.relative_file_path, root_path).to_owned()
            }
            MatchResult::AnalysisLog(log) => {
                log.file = normalize_path_in_project(&log.file, root_path).to_owned()
            }
        }
    }

    pub(crate) fn file_to_match_result(
        file: &Vector<&FileOwner>,
        language: &TargetLanguage,
    ) -> Result<Option<MatchResult>> {
        if file.is_empty() {
            bail!("cannot have file with no versions")
        } else if file.len() == 1 {
            let file = file.last().unwrap();
            if file.new {
                return Ok(Some(MatchResult::CreateFile(CreateFile::file_to_create(
                    file.name.to_string_lossy().as_ref(),
                    &file.source,
                ))));
            } else if let Some(ranges) = &file.matches.borrow().input_matches {
                if ranges.suppressed {
                    return Ok(None);
                }
                return Ok(Some(MatchResult::Match(Match::file_to_match(
                    ranges,
                    &file.source,
                    file.name.to_string_lossy().as_ref(),
                    &file.tree,
                    language,
                ))));
            } else {
                return Ok(None);
            }
        } else {
            return Ok(Some(MatchResult::Rewrite(Rewrite::file_to_rewrite(
                file.front().unwrap(),
                file.back().unwrap(),
                language,
            )?)));
        }
    }

    pub fn file_name(&self) -> Option<&str> {
        match self {
            MatchResult::PatternInfo(pi) => Some(&pi.source_file),
            MatchResult::AllDone(_) => None,
            MatchResult::Match(m) => Some(&m.source_file),
            MatchResult::InputFile(input_file) => Some(&input_file.source_file),
            MatchResult::Rewrite(r) => Some(r.file_name()),
            MatchResult::CreateFile(cf) => Some(cf.file_name()),
            MatchResult::RemoveFile(rf) => Some(rf.file_name()),
            MatchResult::DoneFile(df) => Some(&df.relative_file_path),
            MatchResult::AnalysisLog(log) => Some(&log.file),
        }
    }

    fn extract_original_match(&self) -> Option<&Match> {
        match self {
            MatchResult::DoneFile(_)
            | MatchResult::AnalysisLog(_)
            | MatchResult::InputFile(_)
            | MatchResult::CreateFile(_)
            | MatchResult::AllDone(_)
            | MatchResult::PatternInfo(_) => None,
            MatchResult::Match(m)
            | MatchResult::RemoveFile(RemoveFile { original: m, .. })
            | MatchResult::Rewrite(Rewrite { original: m, .. }) => Some(m),
        }
    }

    /// Extract the original path, if any
    pub fn extract_original_path(&self) -> Option<&str> {
        let original_match = self.extract_original_match()?;
        Some(&original_match.source_file)
    }

    /// Given a MatchResult, create a MatchResult::Rewrite that suppresses the match.
    /// Returns None if it has any issues.
    pub fn get_rewrite_to_suppress(
        &self,
        language: &TargetLanguage,
        pattern_name: Option<&str>,
    ) -> Option<MatchResult> {
        let comment = make_suppress_comment(pattern_name, language);
        let ranges_starts = fs::extract_ranges(self)?
            .iter()
            .map(|r| r.start_byte as usize)
            .collect();

        let original_file_name = self.extract_original_path()?;
        let original_match = self.extract_original_match()?;

        let original_src = std::fs::read_to_string(original_file_name).ok()?;
        let rewritten_content =
            split_string_at_indices(&original_src, ranges_starts).join(&comment);
        let ef = EntireFile::file_to_entire_file(original_file_name, &rewritten_content, None);
        Some(MatchResult::Rewrite(Rewrite::new(
            original_match.clone(),
            ef,
        )))
    }
}

pub fn make_suppress_comment(pattern_name: Option<&str>, language: &TargetLanguage) -> String {
    match pattern_name {
        None => language.make_single_line_comment("grit-ignore"),
        Some(pattern_name) => {
            language.make_single_line_comment(&format!("grit-ignore {pattern_name}"))
        }
    }
}

fn split_string_at_indices(s: &str, indices: Vec<usize>) -> Vec<&str> {
    let mut result = Vec::new();
    let mut prev = 0;
    for &index in indices.iter() {
        if let Some(substring) = s.get(prev..index) {
            result.push(substring);
        }
        prev = index;
    }
    if let Some(substring) = s.get(prev..) {
        result.push(substring);
    }
    result
}

pub fn is_match(result: &MatchResult) -> bool {
    match result {
        MatchResult::AnalysisLog(_) => false,
        MatchResult::Match(_) => true,
        MatchResult::InputFile(_) => false,
        MatchResult::CreateFile(_) => true,
        MatchResult::RemoveFile(_) => true,
        MatchResult::Rewrite(_) => true,
        MatchResult::DoneFile(_) => false,
        MatchResult::AllDone(_) => false,
        MatchResult::PatternInfo(_) => false,
    }
}

pub trait FileMatchResult {
    fn file_name(&self) -> &str;
    fn ranges(&mut self) -> &Vec<Range>;
    // A verb representing the action taken on the file
    fn action() -> &'static str;
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct PatternInfo {
    pub messages: Vec<Message>,
    pub variables: Vec<VariableMatch>,
    pub source_file: String,
    pub parsed_pattern: String,
    pub valid: bool,
}

impl PatternInfo {
    pub fn from_compiled(compiled: Problem, source_file: String) -> Self {
        let node = compiled.tree.root_node();
        let parsed_pattern =
            to_string_pretty(&tree_sitter_node_to_json(&node, &source_file, None)).unwrap();
        Self {
            messages: vec![],
            variables: compiled.compiled_vars(),
            source_file,
            parsed_pattern,
            valid: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct InputFile {
    pub source_file: String,
    pub syntax_tree: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct Match {
    pub messages: Vec<Message>,
    pub variables: Vec<VariableMatch>,
    pub source_file: String,
    pub ranges: Vec<Range>,
    pub debug: String,
}

impl Match {
    fn file_to_match(
        match_ranges: &InputRanges,
        file: &str,
        name: &str,
        tree: &Tree,
        language: &TargetLanguage,
    ) -> Self {
        let input_file_debug_text = to_string_pretty(&tree_sitter_node_to_json(
            &tree.root_node(),
            file,
            Some(language),
        ))
        .unwrap();
        Self {
            debug: input_file_debug_text,
            source_file: name.to_owned(),
            ranges: match_ranges.ranges.to_owned(),
            variables: match_ranges.variables.to_owned(),
            messages: vec![],
        }
    }
}

impl FileMatchResult for Match {
    fn file_name(&self) -> &str {
        &self.source_file
    }
    fn ranges(&mut self) -> &Vec<Range> {
        &self.ranges
    }
    fn action() -> &'static str {
        "matched"
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct EntireFile {
    pub messages: Vec<Message>,
    pub variables: Vec<VariableMatch>,
    pub source_file: String,
    pub content: String,
    pub byte_ranges: Option<Vec<ByteRange>>,
}

impl EntireFile {
    fn file_to_entire_file(name: &str, body: &str, byte_range: Option<&Vec<ByteRange>>) -> Self {
        Self {
            source_file: name.to_owned(),
            content: body.to_owned(),
            variables: vec![],
            messages: vec![],
            byte_ranges: byte_range.map(|r| r.to_owned()),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct Rewrite {
    pub original: Match,
    pub rewritten: EntireFile,
    pub ansi_summary: String,
    pub reason: Option<RewriteReason>,
}

impl From<Rewrite> for MatchResult {
    fn from(r: Rewrite) -> Self {
        MatchResult::Rewrite(r)
    }
}

impl Rewrite {
    fn file_to_rewrite(
        initial: &FileOwner,
        rewrite: &FileOwner,
        language: &TargetLanguage,
    ) -> Result<Self> {
        let original = if let Some(ranges) = &initial.matches.borrow().input_matches {
            Match::file_to_match(
                ranges,
                &initial.source,
                initial.name.to_string_lossy().as_ref(),
                &initial.tree,
                language,
            )
        } else {
            bail!("cannot have rewrite without matches")
        };
        let rewritten = EntireFile::file_to_entire_file(
            rewrite.name.to_string_lossy().as_ref(),
            &rewrite.source,
            rewrite.matches.borrow().byte_ranges.as_ref(),
        );
        Ok(Rewrite::new(original, rewritten))
    }
}

impl FileMatchResult for Rewrite {
    fn file_name(&self) -> &str {
        &self.rewritten.source_file
    }
    fn ranges(&mut self) -> &Vec<Range> {
        &self.original.ranges
    }
    fn action() -> &'static str {
        "rewritten"
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct RewriteReason {
    pub metadata_json: Option<String>,
    pub source: RewriteSource,
    pub name: Option<String>,
    pub level: Option<EnforcementLevel>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
#[serde(rename_all = "camelCase")]
pub enum EnforcementLevel {
    None = 0,
    #[default]
    Info = 1,
    Warn = 2,
    Error = 3,
}

impl FromStr for EnforcementLevel {
    type Err = anyhow::Error;
    fn from_str(s: &str) -> Result<Self> {
        match s {
            "none" => Ok(EnforcementLevel::None),
            "info" => Ok(EnforcementLevel::Info),
            "warn" => Ok(EnforcementLevel::Warn),
            "error" => Ok(EnforcementLevel::Error),
            _ => bail!("'{}' is not a valid level", s),
        }
    }
}

impl fmt::Display for EnforcementLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            EnforcementLevel::None => write!(f, "none"),
            EnforcementLevel::Info => write!(f, "info"),
            EnforcementLevel::Warn => write!(f, "warning"),
            EnforcementLevel::Error => write!(f, "error"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "UPPERCASE")]
pub enum RewriteSource {
    Agent,
    Gritql,
    Stdlib,
    Unknown,
}

impl Rewrite {
    pub fn new(original: Match, rewritten: EntireFile) -> Self {
        Self {
            original,
            rewritten,
            ansi_summary: String::new(),
            reason: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct CreateFile {
    pub rewritten: EntireFile,
    pub ansi_summary: String,
    range: Option<Vec<Range>>,
}

impl From<CreateFile> for MatchResult {
    fn from(r: CreateFile) -> Self {
        MatchResult::CreateFile(r)
    }
}

impl CreateFile {
    fn file_to_create(name: &str, body: &str) -> CreateFile {
        CreateFile {
            rewritten: EntireFile::file_to_entire_file(name, body, None),
            ansi_summary: String::new(),
            range: None,
        }
    }
}

impl FileMatchResult for CreateFile {
    fn file_name(&self) -> &str {
        &self.rewritten.source_file
    }
    fn ranges(&mut self) -> &Vec<Range> {
        match self.range {
            Some(ref r) => r,
            None => {
                let start = Position::first();
                let end = Position::last(&self.rewritten.content);
                self.range = Some(vec![Range::new(
                    start,
                    end,
                    0,
                    self.rewritten.content.len() as u32,
                )]);
                self.ranges()
            }
        }
    }
    fn action() -> &'static str {
        "created"
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct RemoveFile {
    pub original: Match,
    pub ansi_summary: String,
}

impl From<RemoveFile> for MatchResult {
    fn from(r: RemoveFile) -> Self {
        MatchResult::RemoveFile(r)
    }
}

impl FileMatchResult for RemoveFile {
    fn file_name(&self) -> &str {
        &self.original.source_file
    }
    fn ranges(&mut self) -> &Vec<Range> {
        &self.original.ranges
    }
    fn action() -> &'static str {
        "removed"
    }
}

#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct DoneFile {
    pub relative_file_path: String,
    #[serde(skip_serializing)]
    pub has_results: Option<bool>,
    #[serde(skip_serializing)]
    pub file_hash: Option<[u8; 32]>,
    #[serde(skip_serializing)]
    pub from_cache: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    pub message: String,
    pub range: Vec<Range>,
    pub variable_runtime_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct AllDone {
    pub processed: i32,
    pub found: i32,
    pub reason: AllDoneReason,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub enum AllDoneReason {
    NoInputPaths,
    AllMatchesFound,
    MaxResultsReached,
    Aborted,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct ByteRange {
    pub start: usize,
    pub end: usize,
}

impl From<StdRange<usize>> for ByteRange {
    fn from(range: StdRange<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisLog {
    pub level: u16,
    pub message: String,
    pub position: Position,
    pub file: String,
    pub engine_id: String,
    pub range: Option<Range>,
    pub syntax_tree: Option<String>,
    pub source: Option<String>,
}

impl AnalysisLog {
    pub(crate) fn new_error(message: String, file: &str) -> Self {
        Self {
            level: 280,
            message,
            position: Position::first(),
            file: file.to_owned(),
            engine_id: "marzano".to_string(),
            range: None,
            syntax_tree: None,
            source: None,
        }
    }
}

impl From<MarzanoAnalysisLog> for AnalysisLog {
    fn from(log: MarzanoAnalysisLog) -> Self {
        Self {
            level: log.level.unwrap_or(280),
            message: log.message,
            position: log.position.unwrap_or_else(Position::first),
            file: log.file.unwrap_or_default(),
            engine_id: log.engine_id.unwrap_or_else(|| "marzano".to_string()),
            range: log.range,
            syntax_tree: log.syntax_tree,
            source: log.source,
        }
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Serialize, Debug)]
#[serde(rename_all = "lowercase")]
pub enum AnalysisLogLevel {
    Error = 200,
    Warn = 300,
    Info = 400,
    Debug = 500,
}

pub fn derive_log_level(log: &AnalysisLog) -> AnalysisLogLevel {
    match log.level {
        0..=299 => AnalysisLogLevel::Error,
        300..=399 => AnalysisLogLevel::Warn,
        400..=499 => AnalysisLogLevel::Info,
        _ => AnalysisLogLevel::Debug,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_path_in_project_no_root() {
        let path = "./src/main.rs";
        assert_eq!(normalize_path_in_project(path, None), "src/main.rs");
    }

    #[test]
    fn test_normalize_path_in_project_with_root() {
        let path = "/home/user/project/src/main.rs";
        let root_path = PathBuf::from("/home/user/project/");
        assert_eq!(
            normalize_path_in_project(path, Some(&root_path)),
            "src/main.rs"
        );
    }

    #[test]
    #[should_panic]
    fn test_normalize_path_in_project_with_root_no_slash() {
        let path = "/home/user/project/src/main.rs";
        let root_path = PathBuf::from("/home/user/project");
        normalize_path_in_project(path, Some(&root_path));
    }

    #[test]
    fn test_normalize_path_in_project_already_normalized() {
        let path = "src/main.rs";
        assert_eq!(normalize_path_in_project(path, None), "src/main.rs");
    }
}
