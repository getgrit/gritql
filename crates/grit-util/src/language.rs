use crate::{constants::*, traverse, AstNode, CodeRange, Order, Range};
use regex::Regex;

pub enum GritMetaValue {
    Underscore,
    Dots,
    Variable(String),
}

pub trait Language: Sized {
    type Node<'a>: AstNode;

    fn language_name(&self) -> &'static str;

    fn snippet_context_strings(&self) -> &[(&'static str, &'static str)];

    fn alternate_metavariable_kinds(&self) -> &[&'static str] {
        &[]
    }

    fn metavariable_prefix(&self) -> &'static str {
        "$"
    }

    fn comment_prefix(&self) -> &'static str {
        "//"
    }

    fn metavariable_prefix_substitute(&self) -> &'static str {
        "Âµ"
    }

    fn metavariable_regex(&self) -> &'static Regex {
        &VARIABLE_REGEX
    }

    fn replaced_metavariable_regex(&self) -> &'static Regex {
        &REPLACED_VARIABLE_REGEX
    }

    fn metavariable_bracket_regex(&self) -> &'static Regex {
        &BRACKET_VAR_REGEX
    }

    fn exact_variable_regex(&self) -> &'static Regex {
        &EXACT_VARIABLE_REGEX
    }

    fn exact_replaced_variable_regex(&self) -> &'static Regex {
        &EXACT_REPLACED_VARIABLE_REGEX
    }

    fn is_comment(&self, node: &Self::Node<'_>) -> bool;

    fn is_metavariable(&self, node: &Self::Node<'_>) -> bool;

    #[allow(unused_variables)]
    fn is_statement(&self, node: &Self::Node<'_>) -> bool {
        false
    }

    // assumes trim doesn't do anything otherwise range is off
    fn comment_text_range(&self, node: &Self::Node<'_>) -> Option<Range> {
        Some(node.range())
    }

    // in languages we pad such as python or yaml there are
    // some kinds of nodes we don't want to pad, such as python strings.
    // this function identifies those nodes.
    #[allow(unused_variables)]
    fn should_skip_padding(&self, node: &Self::Node<'_>) -> bool {
        false
    }

    #[allow(unused_variables)]
    fn get_skip_padding_ranges_for_snippet(&self, snippet: &str) -> Vec<CodeRange> {
        Vec::new()
    }

    #[allow(unused_variables)]
    fn get_skip_padding_ranges(&self, node: &Self::Node<'_>) -> Vec<CodeRange> {
        let mut ranges = Vec::new();
        for n in traverse(node.walk(), Order::Pre) {
            if self.should_skip_padding(&n) {
                ranges.push(n.code_range())
            }
        }
        ranges
    }

    fn substitute_metavariable_prefix(&self, src: &str) -> String {
        self.metavariable_regex()
            .replace_all(
                src,
                format!("{}$1", self.metavariable_prefix_substitute()).as_str(),
            )
            .to_string()
    }

    fn snippet_metavariable_to_grit_metavariable(&self, src: &str) -> Option<GritMetaValue> {
        src.trim()
            .strip_prefix(self.metavariable_prefix_substitute())
            .map(|s| match s {
                "_" => GritMetaValue::Underscore,
                "..." => GritMetaValue::Dots,
                _ => {
                    let mut s = s.to_owned();
                    s.insert_str(0, self.metavariable_prefix());
                    GritMetaValue::Variable(s)
                }
            })
    }

    /// Check for nodes that should be removed or replaced.
    ///
    /// This is used to "repair" the program after rewriting, such as by
    /// deleting orphaned ranges (like a variable declaration without any
    /// variables). If the node should be removed, it adds a range with a `None`
    /// value. If the node should be replaced, it adds a range with the
    /// replacement value.
    #[allow(unused_variables)]
    fn check_replacements(&self, node: Self::Node<'_>, replacements: &mut Vec<Replacement>) {}

    #[allow(unused_variables)]
    fn take_padding(&self, current: char, next: Option<char>) -> Option<char> {
        if current.is_whitespace() {
            Some(current)
        } else {
            None
        }
    }

    /// Whether snippets should be padded.
    ///
    /// This is generally `true` for languages with relevant whitespace.
    fn should_pad_snippet(&self) -> bool {
        false
    }

    fn make_single_line_comment(&self, text: &str) -> String {
        format!("// {text}\n")
    }
}

#[derive(Clone, Debug)]
pub struct Replacement {
    pub range: Range,
    pub replacement: &'static str,
}

impl Replacement {
    pub fn new(range: Range, replacement: &'static str) -> Self {
        Self { range, replacement }
    }
}
