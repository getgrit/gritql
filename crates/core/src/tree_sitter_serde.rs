use marzano_language::language::{Field, NodeTypes};
use serde_json::{json, Value};
use tree_sitter::Node;

/**
 * Converts a tree-sitter node to a Serde JSON value.
 */
// Todo language should not be an optional
// node types should be it's own trait that Language implements
pub fn tree_sitter_node_to_json(
    node: &Node,
    source: &str,
    language: &impl NodeTypes,
) -> serde_json::Value {
    let sort_id = node.kind_id();
    let node_types = language.node_types();
    let empty: Vec<Field> = vec![];
    let node_fields = if node.is_error() {
        &empty
    } else {
        &node_types[sort_id as usize]
    };
    let mut cursor = node.walk();
    let mut node_json = json!({});
    node_json["SORT"] = node.kind().to_string().into();
    let start_position = json!({
        "line": node.start_position().row() + 1,
        "column": node.start_position().column() + 1,
    });
    let end_position = json!({
        "line": node.end_position().row() + 1,
        "column": node.end_position().column() + 1,
    });
    node_json["RANGE"] = json!({
        "start": start_position,
        "end": end_position,
    });

    if node_fields.is_empty() {
        node_json["TEXT"] = node.utf8_text(source.as_bytes()).unwrap().into();
    }

    let mut processed_child_node_ids: Vec<usize> = vec![];
    for field in node_fields {
        let children_for_field: Vec<Node> =
            node.children_by_field_id(field.id(), &mut cursor).collect();
        if children_for_field.is_empty() {
            continue;
        }
        let child_json = if field.multiple() {
            let mut children_json = vec![];
            for child in children_for_field {
                processed_child_node_ids.push(child.id());
                let child_json = tree_sitter_node_to_json(&child, source, language);
                children_json.push(child_json);
            }
            serde_json::Value::Array(children_json)
        } else {
            assert!(children_for_field.len() == 1);
            let child = &children_for_field[0];
            processed_child_node_ids.push(child.id());
            tree_sitter_node_to_json(child, source, language)
        };
        node_json[field.name()] = child_json;
    }
    let mut unprocessed_child_node: Vec<Value> = vec![];
    for child in node.children(&mut cursor) {
        if !processed_child_node_ids.contains(&child.id()) {
            let child_json = tree_sitter_node_to_json(&child, source, language);
            unprocessed_child_node.push(child_json);
        }
    }
    if !unprocessed_child_node.is_empty() {
        node_json["CHILDREN"] = serde_json::Value::Array(unprocessed_child_node);
    }
    node_json
}
