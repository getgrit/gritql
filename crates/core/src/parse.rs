use crate::api::InputFile;
use anyhow::Result;
use marzano_language::target_language::TargetLanguage;
use std::path::Path;
use tree_sitter::Parser;

#[cfg(feature = "grit-parser")]
pub fn parse_input_file(lang: &TargetLanguage, input: &str, path: &Path) -> Result<InputFile> {
    use crate::tree_sitter_serde::tree_sitter_node_to_json;
    use anyhow::Context;
    use marzano_language::language::Language;
    use serde_json::to_string_pretty;

    let mut parser = Parser::new().context("Failed to create new parser")?;
    parser
        .set_language(lang.get_ts_language())
        .context("Failed to set language for parser")?;
    let tree = parser
        .parse(input.as_bytes(), None)
        .context("Failed to parse input")?
        .context("Parsed tree is empty")?;
    let input_file_debug_text = to_string_pretty(&tree_sitter_node_to_json(
        &tree.root_node(),
        input,
        Some(lang),
    ))
    .context("Failed to convert tree to pretty JSON string")?;
    Ok(InputFile {
        source_file: path.to_string_lossy().to_string(),
        syntax_tree: input_file_debug_text,
    })
}
#[cfg(not(feature = "grit-parser"))]
pub fn parse_input_file(_lang: &TargetLanguage, _input: &str, _path: &Path) -> Result<InputFile> {
    use anyhow::anyhow;

    Err(anyhow!(
        "enable grit-parser feature flag to parse a grit file"
    ))
}

#[cfg(feature = "grit-parser")]
pub fn make_grit_parser() -> Result<Parser> {
    use anyhow::Context;

    let mut parser = Parser::new().unwrap();
    parser
        .set_language(&tree_sitter_gritql::language().into())
        .with_context(|| "Failed to load grit language")?;
    Ok(parser)
}

#[cfg(not(feature = "grit-parser"))]
pub fn make_grit_parser() -> Result<Parser> {
    use anyhow::anyhow;

    Err(anyhow!(
        "enable grit-parser feature flag to make a grit parser"
    ))
}
