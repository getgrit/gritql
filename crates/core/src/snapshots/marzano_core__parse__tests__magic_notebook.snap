---
source: crates/core/src/parse.rs
expression: code
---
module                         | x = 1\npip install maybeshow\ncolors = 6\nprint(colors)\ndef maybeshow():\n    pass\n\n%timeit foo()\ndef foo():\n    pass\n%timeit other('hello maybeshow')\ndef sample_func(xx):\n    """\n    转置 (transpose)\n    """\n    return xx.T\n# https://github.com/astral-sh/ruff-vscode/issues/362\nDEFAULT_SYSTEM_PROMPT = (\n    "Ты — Сайга, русскоязычный автоматический ассистент. "\n    "Ты разговариваешь с людьми и помогаешь им."\n)\n%%timeit\nprint('hello maybeshow')\n
assignment                     | x = 1
identifier                     | x
=                              | =
integer                        | 1
ERROR                          | pip install
identifier                     | pip
identifier                     | maybeshow
assignment                     | colors = 6
identifier                     | colors
=                              | =
integer                        | 6
call                           | print(colors)
identifier                     | print
argument_list                  | (colors)
(                              | (
identifier                     | colors
)                              | )
function_definition            | def maybeshow():\n    pass
def                            | def
identifier                     | maybeshow
(                              | (
)                              | )
:                              | :
block                          | pass
pass_statement                 | pass
pass                           | pass
ERROR                          | %timeit
modulo                         | %
identifier                     | timeit
call                           | foo()
identifier                     | foo
argument_list                  | ()
(                              | (
)                              | )
function_definition            | def foo():\n    pass
def                            | def
identifier                     | foo
(                              | (
)                              | )
:                              | :
block                          | pass
pass_statement                 | pass
pass                           | pass
ERROR                          | %timeit
modulo                         | %
identifier                     | timeit
call                           | other('hello maybeshow')
identifier                     | other
argument_list                  | ('hello maybeshow')
(                              | (
string                         | 'hello maybeshow'
string_start                   | '
string_content                 | hello maybeshow
string_end                     | '
)                              | )
function_definition            | def sample_func(xx):\n    """\n    转置 (transpose)\n    """\n    return xx.T
def                            | def
identifier                     | sample_func
(                              | (
identifier                     | xx
)                              | )
:                              | :
block                          | """\n    转置 (transpose)\n    """\n    return xx.T
string                         | """\n    转置 (transpose)\n    """
string_start                   | """
string_content                 | \n    转置 (transpose)\n    
string_end                     | """
return_statement               | return xx.T
return                         | return
attribute                      | xx.T
identifier                     | xx
.                              | .
identifier                     | T
comment                        | # https://github.com/astral-sh/ruff-vscode/issues/362
assignment                     | DEFAULT_SYSTEM_PROMPT = (\n    "Ты — Сайга, русскоязычный автоматический ассистент. "\n    "Ты разговариваешь с людьми и помогаешь им."\n)
identifier                     | DEFAULT_SYSTEM_PROMPT
=                              | =
parenthesized_expression       | (\n    "Ты — Сайга, русскоязычный автоматический ассистент. "\n    "Ты разговариваешь с людьми и помогаешь им."\n)
(                              | (
concatenated_string            | "Ты — Сайга, русскоязычный автоматический ассистент. "\n    "Ты разговариваешь с людьми и помогаешь им."
string                         | "Ты — Сайга, русскоязычный автоматический ассистент. "
string_start                   | "
string_content                 | Ты — Сайга, русскоязычный автоматический ассистент. 
string_end                     | "
string                         | "Ты разговариваешь с людьми и помогаешь им."
string_start                   | "
string_content                 | Ты разговариваешь с людьми и помогаешь им.
string_end                     | "
)                              | )
ERROR                          | %%
modulo                         | %
modulo                         | %
identifier                     | timeit
call                           | print('hello maybeshow')
identifier                     | print
argument_list                  | ('hello maybeshow')
(                              | (
string                         | 'hello maybeshow'
string_start                   | '
string_content                 | hello maybeshow
string_end                     | '
)                              | )
